### PR TITLE
BUG: Fix edge cases in merge_asof() by comparing factorized keys (#13709)

### LINKED ISSUE #13709: pd.merge_asof() matches out of tolerance when timestamps are duplicated
This is a continuation of #13695.

Starting with the original DataFrames from that issue:

```
df1 = pd.DataFrame({'time':pd.to_datetime(['2016-07-15 13:30:00.030']), 'username':['bob']})
df2 = pd.DataFrame({'time':pd.to_datetime(['2016-07-15 13:30:00.000', '2016-07-15 13:30:00.030']), 'version':[1, 2]})
```

I now get the null:

```
In [82]: pd.merge_asof(df1, df2, on='time', allow_exact_matches=False, tolerance=pd.Timedelta('10ms'))
Out[82]:
                     time username  version
0 2016-07-15 13:30:00.030      bob      NaN
```

However, if I change the first DataFrame to have duplicate timestamps:

```
df1 = pd.DataFrame({'time':pd.to_datetime(['2016-07-15 13:30:00.030', '2016-07-15 13:30:00.030']), 'username':['bob', 'charlie']})
```

then the bug reappears:

```
In [85]: pd.merge_asof(df1, df2, on='time', allow_exact_matches=False, tolerance=pd.Timedelta('10ms'))
Out[85]:
                     time username  version
0 2016-07-15 13:30:00.030      bob        1
1 2016-07-15 13:30:00.030  charlie        1
```

This is in pandas version 0.18.0+418.gc46dcfa.

### PR DESCRIPTION
**Pull request by chrisaycock**
closes #13709 

Also removes unnecessary check_duplicates.

### CODE REVIEW

**Comment by jreback** [comment]
this is have the same perf characteristics? (I was manually doing the perf comparisons).....prob should have some asv's for this....

File: pandas/src/join.pyx [hunk 1]
```diff
@@ -126,150 +126,56 @@ def left_outer_join(ndarray[int64_t] left, ndarray[int64_t] right,
...
-    right_pos = right_count[0]
-
-    left_indexer = np.empty(count, dtype=np.int64)
-    right_indexer = np.empty(count, dtype=np.int64)
-
-    for i in range(1, max_groups + 1):
-        lc = left_count[i]
-        rc = right_count[i]
-
-        if rc == 0:
```
Comment (jreback) [inline pandas/src/join.pyx]: really simplified!

**Comment by chrisaycock** [comment]
@jreback I had thought about perf issues, but since the logic is a lot smaller (avoids the call to groupsort_indexer(), for example), it has to be faster.

**Comment by jreback** [comment]
actually I suspect it's much slower

the bottleneck never was the cython code
rather the grouping / indexing

that is why the check_duplicates helps

**Comment by chrisaycock** [comment]
Can you show me the code you used for the perf tests? I'll run them manually and report back.

**Comment by jreback** [comment]
notebook is [here](http://nbviewer.jupyter.org/gist/jreback/5f089d308750c89b2a7d7446b790c056)

these used the original trades.csv, quotes.csv that you had sent me:
```
-rw-rw-r--@ 1 jreback  staff  17202707 Jun  2 18:19 trades.csv
-rw-rw-r--@ 1 jreback  staff  47853000 Jun  2 18:19 quotes.csv
```

they are not in the repo; we should prob generate some for test purposes

**Comment by chrisaycock** [comment]
I've refreshed my data sample since that notebook, but my DataFrames are a similar size.

```
In [3]: trades.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 319018 entries, 0 to 319017
Data columns (total 4 columns):
time        319018 non-null datetime64[ns]
ticker      319018 non-null category
price       316669 non-null float64
quantity    319018 non-null int64
dtypes: category(1), datetime64[ns](1), float64(1), int64(1)
memory usage: 7.6 MB

In [4]: quotes.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1032780 entries, 0 to 1032779
Data columns (total 4 columns):
time      1032780 non-null datetime64[ns]
ticker    1032780 non-null category
bid       1030431 non-null float64
ask       1030431 non-null float64
dtypes: category(1), datetime64[ns](1), float64(2)
memory usage: 24.6 MB
```

I ran the numbers with a fresh build both before and after my code change.

Before:
```
In [5]: %timeit -n 1 -r 1 pd.merge_asof(trades, quotes, on='time', by='ticker')
1 loop, best of 1: 970 ms per loop
```

After:
```
In [5]: %timeit -n 1 -r 1 pd.merge_asof(trades, quotes, on='time', by='ticker')
1 loop, best of 1: 835 ms per loop
```

So my code change is moderately faster, which is what I expect. The Python logic didn't really change since the `check_duplicates` is now gone (equivalent to the `False` case). And since the Cython logic was never much of a bottleneck, changes there provide only modest performance improvements.

File: pandas/tools/merge.py [hunk 2]
```diff
@@ -436,7 +427,7 @@ def _merger(x, y):
     if by is not None:
         result, groupby = _groupby_and_merge(by, on, left, right,
                                              lambda x, y: _merger(x, y),
-                                             check_duplicates=check_duplicates)
+                                             check_duplicates=False)
```
Comment (chrisaycock) [inline pandas/tools/merge.py]: Note that `check_duplicates` is always `False` now, since it is no longer needed.

**Comment by jreback** [comment]
try it with duplicates

**Comment by chrisaycock** [comment]
Are you referring to duplicate timestamps and tickers? That's already in my sample data:

```
In [16]: quotes
Out[16]:
                           time ticker     bid     ask
0       2016-07-15 13:30:00.000   GOOG  724.24  725.73
1       2016-07-15 13:30:00.002   MSFT   53.83   53.95
2       2016-07-15 13:30:00.002   MSFT   53.83   53.95
3       2016-07-15 13:30:00.004   GOOG  722.75  725.73
4       2016-07-15 13:30:00.036   GOOG  725.48  725.73
5       2016-07-15 13:30:00.036   AAPL   98.92   99.05
6       2016-07-15 13:30:00.036   GOOG  722.75  725.73
7       2016-07-15 13:30:00.036   AAPL   98.92   99.05
8       2016-07-15 13:30:00.037   MSFT   53.84   53.95
9       2016-07-15 13:30:00.037   MSFT   53.84   53.95
...                         ...    ...     ...     ...
```

And of course the regression tests with duplicate data pass.

**Comment by jreback** [comment]
can u add an asv (prob need to generate data given a seed)

**Comment by chrisaycock** [comment]
Will do.

**Comment by chrisaycock** [comment]
I added a couple of simple benchmarks.

```
· Running 4 total benchmarks (2 commits * 1 environments * 2 benchmarks)
[  0.00%] · For pandas commit hash aab3e3cc:
[  0.00%] ·· Building for conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..............................
[  0.00%] ·· Benchmarking conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 25.00%] ··· Running join_merge.merge_asof_by.time_merge_asof_by                                              539.69ms
[ 50.00%] ··· Running join_merge.merge_asof_noby.time_merge_asof_noby                                           77.95ms
[ 50.00%] · For pandas commit hash 2166ac13:
[ 50.00%] ·· Building for conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
[ 50.00%] ·· Benchmarking conda-py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 75.00%] ··· Running join_merge.merge_asof_by.time_merge_asof_by                                              708.91ms
[100.00%] ··· Running join_merge.merge_asof_noby.time_merge_asof_noby                                           97.48ms
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

File: asv_bench/benchmarks/join_merge.py [hunk 3]
```diff
@@ -293,6 +293,43 @@ def time_join_dataframe_integer_key(self):
         merge(self.df, self.df2, on='key1')
 
 
+class merge_asof_noby(object):
+
+    def setup(self):
+        np.random.seed(0)
+        one_count = 200000
+        two_count = 1000000
+        self.df1 = pd.DataFrame({'time': np.random.random_integers(0, one_count/20, one_count),
```
Comment (jreback) [inline asv_bench/benchmarks/join_merge.py]: use ``np.random.randint(....)`` as ``random_integers`` is deprecated

File: asv_bench/benchmarks/join_merge.py [hunk 4]
```diff
@@ -293,6 +293,43 @@ def time_join_dataframe_integer_key(self):
...
+                                 'value1': np.random.randn(one_count)})
+        self.df2 = pd.DataFrame({'time': np.random.random_integers(0, two_count/20, two_count),
+                                 'value2': np.random.randn(two_count)})
+        self.df1.sort_values('time', inplace=True)
```
Comment (jreback) [inline asv_bench/benchmarks/join_merge.py]: more idiomatic is just put the ``.sort_values(...)`` w/o inplace

**Comment by jreback** [comment]
@chrisaycock lgtm. just a very small change in asv. ping when green.

**Comment by chrisaycock** [comment]
Updated as per your notes. Thanks.

**Comment by chrisaycock** [comment]
@jreback It's green!

**Comment by jreback** [comment]
thank you sir!

**Pull request merged**